projects/output-and-error-handling/client/main.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+// Comments ----------
+// I write a lot of comments, it has helped me the entire course to learn, understand and remember
+// They can all be removed at the end to clean it up
+
+// Error Messages ----------
+// The Errors are deliberately numbered [x] so I could follow them up the call stack
+// They can be replaced with "user friendly" messages later
+
+func parseResponseBody(response *http.Response) (string, error) {
+	// check if there is a response body
+	if response.Body == nil {
+		return "", fmt.Errorf("[2] there is no response body")
+	}
+
+	// try to read the response body
+	responseBody, err := io.ReadAll(response.Body)
+
+	// if we cannot read the response body then return an error
+	if err != nil {
+		return "", fmt.Errorf("[2] failed to read response body : %w", err)
+	}
+
+	return string(responseBody), nil
+}
+
+func parseRetryAfterHeader(retryAfterHeaderString string) (time.Duration, error) {
+	// fmt.Println("--- DEBUG retryAfterHeaderString", retryAfterHeaderString)
+
+	// first condition "a while"
+	if retryAfterHeaderString == "a while" {
+		return 0, fmt.Errorf("[3] retry-after header is 'a while'")
+	}
+
+	// second condition an integer in string format
+	retryAfterInteger, err := strconv.Atoi(retryAfterHeaderString)
+
+	// if we successfully parse an integer
+	if err == nil {
+		// convert the integer into seconds and return it
+		retryDuration := time.Duration(retryAfterInteger) * time.Second
+		return retryDuration, nil
+	}
+
+	// third condition a timestamp in string format
+	// (!) http.ParseTime can parse various formats
+	// specified in the RFCs for HTTP (RFC1123, RFC850, ANSI C etc)
+	// so it is a better choice than time.Parse(RFC1123, _)
+	retryAfterTimeStamp, err := http.ParseTime(retryAfterHeaderString)
+
+	// if we cannot parse the timestamp then return an error
+	if err != nil {
+		return 0, fmt.Errorf("[3] retry-after header is invalid: %w", err)
```

Comment:
It may be useful to include the contents of the header the error message so that someone could debug _why_ it was invalid (and maybe report a bug or fix the server or client)